### PR TITLE
CMake build for ip2

### PR DESCRIPTION
**Pull request by kgerheiser**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+if(NOT CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel|GNU|Clang|AppleClang)$")
+  message(WARNING "Compiler not officially supported: ${CMAKE_Fortran_COMPILER_ID}")
+endif()
+
+if(OPENMP)
+  find_package(OpenMP REQUIRED COMPONENTS Fortran)
+endif()
+
+find_package(sp CONFIG REQUIRED)
```
Comment (aerorahul) [inline CMakeLists.txt]: Don’t need CONFIG

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  target_compile_definitions(${lib_name} PRIVATE "LSIZE=${kind_definition}")
+  set_target_properties(${lib_name} PROPERTIES COMPILE_FLAGS "${BUILD_FLAGS}")
+  set_target_properties(${lib_name} PROPERTIES Fortran_MODULE_DIRECTORY "${module_dir}")
+
+  if(OpenMP_Fortran_FOUND)
+    target_link_libraries(${lib_name} PUBLIC OpenMP::OpenMP_Fortran)
+  endif()
+
+  target_link_libraries(${lib_name} PRIVATE sp::sp_${kind})
```
Comment (aerorahul) [inline src/CMakeLists.txt]: Shouldn’t this be PUBLIC for transitive to work.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#list(APPEND CMAKE_CTEST_ARGUMENTS "--output-on-failure")
+
+file(COPY data/ DESTINATION data)
+
+find_package(sp REQUIRED)
```
Comment (aerorahul) [inline tests/CMakeLists.txt]: Not needed here. It is covered at root level of repo

**Review by aerorahul — CHANGES_REQUESTED** [review]
Some minor comments. 
Also, we should make use of the version-config file cmake creates to report version

**Review by GeorgeGayno-NOAA — COMMENTED** [review]
Can you please remove the old build system?

**Comment by kgerheiser** [comment]
Is there a version for this library? I put `1.0.0`, because I couldn't find one.

**Comment by GeorgeGayno-NOAA** [comment]
> Is there a version for this library? I put `1.0.0`, because I couldn't find one.

Version 1.0.0 was tagged and implemented on WCOSS a while back.  So we should increase the version number.

**Comment by kgerheiser** [comment]
I bumped the version number and I removed the make build system

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+file(STRINGS "VERSION" pVersion)
+
+project(
+  ip2
+  VERSION ${pVersion}
+  LANGUAGES Fortran)
+
+option(OPENMP "use OpenMP threading" OFF)
+
```
Comment (aerorahul) [inline CMakeLists.txt:12]: Need an option to `ENABLE_TESTS`

File: tests/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#list(APPEND CMAKE_CTEST_ARGUMENTS "--output-on-failure")
+
+file(COPY data/ DESTINATION data)
```
Comment (aerorahul) [inline tests/CMakeLists.txt]: why copy? why not link?

Comment (kgerheiser) [reply tests/CMakeLists.txt]: true

**Review by aerorahul — CHANGES_REQUESTED** [review]
missing option and link instead of copy comment.

**Comment by kgerheiser** [comment]
@GeorgeGayno-NOAA 

I added the updated lat-lon thing. Let me know if this is right.

https://github.com/kgerheiser/NCEPLIBS-ip2/blob/feature/cmake/src/gdswzd_rot_equid_cylind_egrid_mod.f90

**Review by aerorahul — APPROVED** [review]
looks good.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA
> 
> I added the updated lat-lon thing. Let me know if this is right.
> 
> https://github.com/kgerheiser/NCEPLIBS-ip2/blob/feature/cmake/src/gdswzd_rot_equid_cylind_egrid_mod.f90

Looks like you got the correct version.  Before merging I would like to run my unit and regression tests.  Unfortunately, they ceased to work when Dexin added his build system.  So I will use the tests from a previous snap shot.

**Comment by GeorgeGayno-NOAA** [comment]
We store 'iplib.sh' files in the include directory.  Can that file be added as part of the install?

**Comment by kgerheiser** [comment]
You mean `iplib_4.h`, etc  (not `.sh`)?

**Comment by GeorgeGayno-NOAA** [comment]
> You mean `iplib_4.h`, etc (not `.sh`)?

Yes.  And they are all stored with the same name - iplib.h.   On Dell...

```
m71a1@ /gpfs/dell1/nco/ops/nwprod/lib/ip2/v1.0.0/ips/18.0.1\ $ ls -lR include/*/*.h
-rwxr-xr-x 1 nwprod prod 325 Aug  2  2019 include/ip2_v1.0.0_4/iplib.h
-rwxr-xr-x 1 nwprod prod 343 Aug  2  2019 include/ip2_v1.0.0_8/iplib.h
-rwxr-xr-x 1 nwprod prod 337 Aug  2  2019 include/ip2_v1.0.0_d/iplib.h

```

**Comment by kgerheiser** [comment]
Ok, that's easy.

**Comment by kgerheiser** [comment]
Done

It installs `iplib_${kind}` as `iplib.h` for each kind

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,79 @@
+include("list_of_files.cmake")
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "-g -auto -convert big_endian -assume byterecl -fp-model strict -fpp")
+  set(fortran_d_flags "-i4 -r8")
```
Comment (aerorahul) [inline src/CMakeLists.txt]: Don’t need ‘ i4’

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  target_link_libraries(${lib_name} PUBLIC sp::sp_${kind})
+  
+  target_include_directories(${lib_name}
+    PUBLIC $<BUILD_INTERFACE:${module_dir}>
+    $<INSTALL_INTERFACE:include_${kind}>)
+
+  list(APPEND LIB_TARGETS ${lib_name})
+  
+  install(DIRECTORY ${module_dir} DESTINATION ${CMAKE_INSTALL_PREFIX})
+  install(FILES iplib_${kind}.h DESTINATION ${CMAKE_INSTALL_PREFIX}/include_${kind} RENAME iplib.h)
```
Comment (aerorahul) [inline src/CMakeLists.txt]: If this library is called ‘ip2’, why is the header ‘iplib.h’?

**Review by aerorahul — CHANGES_REQUESTED** [review]
How different really is ip2 from up and if not very, why are they not a single library?
I see a lot of repeated code between ip and ip2.

**Comment by GeorgeGayno-NOAA** [comment]
> How different really is ip2 from up and if not very, why are they not a single library?
> I see a lot of repeated code between ip and ip2.

They are very similar.  ip2 does its geo-references using the grib2 grid definition template, while ip uses the grib1 template.    The grib2 template has more information.  But in principle they could be combined.

File: src/CMakeLists.txt (hunk 7 above)
Comment (GeorgeGayno-NOAA) [reply src/CMakeLists.txt]: That is a good point.  ip2lib was created from iplib.  I must have forgot to change that.  @kgerheiser can we rename them?

**Comment by GeorgeGayno-NOAA** [comment]
Should threading be turned on by default?  That is how we normally compile ip/ip2.

**Comment by GeorgeGayno-NOAA** [comment]
Checked out 5d6821f on Dell (which contains the last fully working version of the unit and regression tests).  Ran all tests using Kyle's branch at 001e181.  All tests passed.  Yeah!

**Comment by GeorgeGayno-NOAA** [comment]
@kgerheiser  At some point, we need to get the unit and regression tests working.  I don't know if you want to do that with this pull request or wait.

**Comment by kgerheiser** [comment]
Fixed all those things. Good to merge then?

File: CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+option(OPENMP "use OpenMP threading" OFF)
+option(ENABLE_TESTS "Build pfunit tests?" OFF)
+
+if(NOT CMAKE_BUILD_TYPE MATCHES "^(Debug|Release|RelWithDebInfo|MinSizeRel)$")
+  message(STATUS "Setting build type to 'Release' as none was specified.")
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the type of build." FORCE)
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+endif()
+
+if(NOT CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel|GNU|Clang|AppleClang)$")
```
Comment (aerorahul) [inline CMakeLists.txt]: Fortran does not have `Clang|AppleClang`.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,79 @@
+include("list_of_files.cmake")
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "-g -auto -convert big_endian -assume byterecl -fp-model strict -fpp")
+  set(fortran_d_flags "-r8")
+  set(fortran_8_flags "-i8 -r8")
+elseif(CMAKE_Fortran_COMPILER_ID MATCHES "^(GNU|Clang|AppleClang)$")
```
Comment (aerorahul) [inline src/CMakeLists.txt]: No `Clang|AppleClang` for Fortran.

File: tests/README [hunk 10]
```diff
@@ -0,0 +1,204 @@
+
```
Comment (aerorahul) [inline tests/README]: Can you turn this into `README.md`
Markdown files are so much better to read.

**Review by aerorahul — APPROVED** [review]
looks good.

**Pull request merged**